lib/srv/app/gcp/handler.go

```diff
@@ -0,0 +1,276 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gcp
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"cloud.google.com/go/iam/credentials/apiv1/credentialspb"
+	"github.com/gravitational/oxy/forward"
+	oxyutils "github.com/gravitational/oxy/utils"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/api/utils/gcp"
+	"github.com/gravitational/teleport/lib/cloud"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/httplib"
+	"github.com/gravitational/teleport/lib/srv/app/common"
+	"github.com/gravitational/teleport/lib/utils"
+	awsutils "github.com/gravitational/teleport/lib/utils/aws"
+)
+
+// HandlerConfig is the configuration for an GCP app-access handler.
+type HandlerConfig struct {
+	// RoundTripper is the underlying transport given to an oxy Forwarder.
+	RoundTripper http.RoundTripper
+	// Log is the Logger.
+	Log logrus.FieldLogger
+	// Clock is used to override time in tests.
+	Clock clockwork.Clock
+
+	// CloudClients is a set of cloud clients that Teleport supports.
+	CloudClients cloud.Clients
+
+	// getAccessToken is a function for getting access token, pluggable for the sake of testing.
+	generateAccessToken generateAccessTokenFunc
+}
+
+// CheckAndSetDefaults validates the HandlerConfig.
+func (s *HandlerConfig) CheckAndSetDefaults() error {
+	if s.RoundTripper == nil {
+		tr, err := defaults.Transport()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		s.RoundTripper = tr
+	}
+	if s.Clock == nil {
+		s.Clock = clockwork.NewRealClock()
+	}
+	if s.Log == nil {
+		s.Log = logrus.WithField(trace.Component, "gcp:fwd")
+	}
+
+	if s.CloudClients == nil {
+		s.CloudClients = cloud.NewClients()
+	}
+
+	if s.generateAccessToken == nil {
+		s.generateAccessToken = s.generateAccessTokenDefaultImpl
+	}
+
+	return nil
+}
+
+type generateAccessTokenFunc func(ctx context.Context, serviceAccount string, scopes []string) (*credentialspb.GenerateAccessTokenResponse, error)
+
+func (s *HandlerConfig) generateAccessTokenDefaultImpl(ctx context.Context, serviceAccount string, scopes []string) (*credentialspb.GenerateAccessTokenResponse, error) {
```

Comment:
Can we remove the `DefaultIImpl` suffix 
```suggestion
func (s *HandlerConfig) generateAccessToken(ctx context.Context, serviceAccount string, scopes []string) (*credentialspb.GenerateAccessTokenResponse, error) {
```

Reply:
That would result in both field and a method with the same name, which isn't allowed.

Reply:
Since I wanted to test `generateAccessToken` (which used to be called `generateAccessTokenDefaultImpl`), I changed the extensibility point used in tests to be the cloud clients. This required a few helper types, but we have better test coverage now.